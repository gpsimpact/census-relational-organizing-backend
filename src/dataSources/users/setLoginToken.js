export default redis => async (userId, token) => {
  await redis.set(token, userId, "EX", 60 * 60 * 24 * 1000);
  return true;
};
