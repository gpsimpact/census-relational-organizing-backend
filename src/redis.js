import Redis from "ioredis";
import { RedisPubSub } from "graphql-redis-subscriptions";
// import RedisSMQ from "rsmq";

export const pubsub = new RedisPubSub({
  publisher: new Redis(process.env.REDIS_URL),
  subscriber: new Redis(process.env.REDIS_URL)
});

// export const rsmq = new RedisSMQ(
//   process.env.REDIS_URL && require("redis-url").parse(process.env.REDIS_URL)
// );

// io redis client
export default new Redis(process.env.REDIS_URL);
