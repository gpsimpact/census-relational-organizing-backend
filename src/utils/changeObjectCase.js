import { mapKeys, camelCase, snakeCase } from "lodash";

export const toCamelCase = obj => {
  if (!obj) return obj;
  return mapKeys(obj, (value, key) => {
    return camelCase(key);
  });
};

export const toSnakeCase = obj =>
  mapKeys(obj, (value, key) => {
    return snakeCase(key);
  });
