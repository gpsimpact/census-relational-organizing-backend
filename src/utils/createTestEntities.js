import faker from "faker";
import { createGDS } from "@jakelowen/sqorn-graphql-filters";
import { sq } from "../db";

export const uuid = () => faker.random.uuid();

export const createTestUser = async (
  name,
  email,
  abbreviation,
  active = true
) =>
  createGDS(sq.from`users`)({
    name: name || `${faker.name.firstName()} ${faker.name.lastName()}`,
    email: email || faker.internet.email().toLowerCase(),
    abbreviation: faker.lorem.word().substring(0, 7),
    active
  });

export const createTestGlobalPerm = async (userId, permission) =>
  createGDS(sq.from`global_permissions`)({
    userId,
    permission
  });

export const createTestClient = async (active = true) =>
  createGDS(sq.from`clients`)({
    name: faker.company.companyName(),
    abbreviation: faker.random.alphaNumeric(7),
    slug: faker.lorem.slug(),
    active
  });

export const createTestOLPermission = async (userId, clientId, permission) =>
  createGDS(sq.from`client_permissions`)({
    clientId,
    userId,
    permission
  });

export const createTestCycle = async (
  clientId,
  cycleName,
  cycleCategory,
  engagementDates
) =>
  createGDS(sq.from`cycles`)({
    clientId,
    name: cycleName || faker.lorem.sentence(),
    category: cycleCategory,
    engagementDates: engagementDates || `[2018-01-01,2018-12-31)`
  });

export const createTestIO = async (
  insertion_order_id,
  client_id,
  created_by,
  attribution,
  campaign_name,
  campaign_description,
  program_phase,
  commission_rate = 0.15
) =>
  createGDS(sq.from`insertion_orders_revisions`)({
    insertion_order_id: insertion_order_id || faker.random.uuid(),
    client_id,
    created_by,
    attribution: attribution || faker.name.firstName(),
    campaign_name: campaign_name || faker.lorem.words(4),
    campaign_description: campaign_description || faker.lorem.word(),
    program_phase: program_phase || faker.lorem.word(),
    commission_rate
  });

export const createTestIOLineItem = async (
  insertion_order_line_item_id,
  insertion_order_id,
  created_by,
  category,
  platform,
  objective,
  cost_gross,
  cost_net,
  active_dates = "[2018-01-01, 2018-12-31)",
  commission_rate = 0.15
) =>
  createGDS(sq.from`insertion_orders_line_items_revisions`)({
    insertion_order_line_item_id:
      insertion_order_line_item_id || faker.random.uuid(),
    insertion_order_id,
    category: category || faker.lorem.word(),
    platform: platform || faker.lorem.word(),
    objective: objective || faker.lorem.word(),
    active_dates,
    cost_gross:
      cost_gross || faker.random.number({ min: 0, max: 100, precision: 2 }),
    cost_net:
      cost_net || faker.random.number({ min: 0, max: 100, precision: 2 }),
    commission_rate,
    created_by
  });
