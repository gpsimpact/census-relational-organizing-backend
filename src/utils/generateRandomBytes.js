import { randomBytes } from "crypto";
import { promisify } from "util";

export default async length => {
  const randomBytesPromiseified = promisify(randomBytes);
  return (await randomBytesPromiseified(length)).toString("hex");
};
