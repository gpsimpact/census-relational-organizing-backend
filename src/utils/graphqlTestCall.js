import { graphql } from "graphql";
import { makeExecutableSchema } from "graphql-tools";
import { applyMiddleware } from "graphql-middleware";
import transformsMiddleware from "../middleware/transforms";
import permissionsMiddleware from "../middleware/permissions";

import makeContext from "../context";
import { default as typeDefs } from "../typeDefinitions";
import resolvers from "../resolvers";

export const schema = makeExecutableSchema({
  typeDefs,
  resolvers
});

const schemaWithMiddleware = applyMiddleware(
  schema,
  transformsMiddleware,
  permissionsMiddleware
);

export const makeTestContext = (userId, contextOverRides) => {
  const context = makeContext(
    {
      session: {
        userId,
        destroy: () => {}
      }
    },
    {
      clearCookie: () => {}
    }
  );
  const testContext = Object.assign(context, contextOverRides);
  return testContext;
};

export const graphqlTestCall = async (
  query,
  variables,
  userId,
  contextOverRides
) => {
  // const schema = await generateSchema();
  const testContext = makeTestContext(userId, contextOverRides);
  return graphql(
    schemaWithMiddleware,
    query,
    undefined,
    testContext,
    variables
  );
};
