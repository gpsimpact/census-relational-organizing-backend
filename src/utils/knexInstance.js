import Knex from "knex";

const knexfile = require("../../knexfile");

export default Knex({
  ...knexfile
});
