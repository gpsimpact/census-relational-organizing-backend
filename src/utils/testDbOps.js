// import { map } from "lodash";
import knex from "./knexInstance";
import redis from "../redis";

// const tableList = [
//   "users",
//   "clients",
//   "global_permissions",
//   "client_permissions",
//   "insertion_orders_current",
//   "insertion_orders_revisions",
//   "insertion_orders_line_items_current",
//   "insertion_orders_line_items_revisions"
// ];

export const dbUp = async () => {
  await knex.raw(`
    TRUNCATE TABLE users CASCADE;
    TRUNCATE TABLE clients CASCADE;
    TRUNCATE TABLE global_permissions CASCADE;
    TRUNCATE TABLE client_permissions CASCADE;
    TRUNCATE TABLE insertion_orders_current CASCADE;
    TRUNCATE TABLE insertion_orders_revisions CASCADE;
    TRUNCATE TABLE insertion_orders_line_items_current CASCADE;
    TRUNCATE TABLE insertion_orders_line_items_revisions CASCADE;
  `);
  await redis.flushdb();
};
// await Promise.all(
//   knex.raw(`
//     TRUNCATE TABLE users CASCADE;
//     TRUNCATE TABLE clients CASCADE;
//     TRUNCATE TABLE global_permissions CASCADE;
//     TRUNCATE TABLE client_permissions CASCADE;
//     TRUNCATE TABLE insertion_orders_current CASCADE;
//     TRUNCATE TABLE insertion_orders_revisions CASCADE;
//     TRUNCATE TABLE insertion_orders_line_items_current CASCADE;
//     TRUNCATE TABLE insertion_orders_line_items_revisions CASCADE;
//   `),
//   redis.flushdb()
// );

export const dbDown = async () => {
  // await Promise.all(
  //   map(tableList, table => knex.raw(`TRUNCATE TABLE ${table} CASCADE;`)),
  //   redis.flushdb()
  // );
};
