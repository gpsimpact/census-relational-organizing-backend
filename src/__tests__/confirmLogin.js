// import faker from "faker";
import { graphqlTestCall } from "../utils/graphqlTestCall";
import { dbUp, dbDown } from "../utils/testDbOps";
import { createTestUser } from "../utils/createTestEntities";
import setLoginToken from "../dataSources/users/setLoginToken";
import redis from "../redis";

const CONFIRM_LOGIN_MUTATION = `
    mutation confirmLogin($token: String!){
      confirmLogin(token: $token) {
            code
            success
            message
            item {
              id
            }
        }
    }
`;

const setToken = setLoginToken(redis);

beforeEach(async () => {
  await dbUp();
});

afterEach(async () => {
  await dbDown();
});

describe("confirmLoginResolver", () => {
  test("Happy Path", async () => {
    // insert record
    const user = await createTestUser();
    const token = "fdsfdsafdsfew23r4ewr432r";
    await setToken(user.id, token);

    const rToken1 = await redis.get(token);
    expect(rToken1).toBe(user.id);

    const response = await graphqlTestCall(CONFIRM_LOGIN_MUTATION, {
      token
    });
    expect(response.data.confirmLogin.code).toBe("OK");
    expect(response.data.confirmLogin.success).toBe(true);
    expect(response.data.confirmLogin.item.id).toEqual(user.id);

    const rToken2 = await redis.get(token);
    expect(rToken2).toBeNull();
  });

  test("throws error if logged in", async () => {
    const user = await createTestUser();
    const token = "fdsfdsafdsfew23r4ewr432r";
    await setToken(user.id, token);

    const res = await graphqlTestCall(
      CONFIRM_LOGIN_MUTATION,
      {
        token
      },
      user.id
    );
    expect(res.errors.length).toBe(1);
    expect(res.errors).toMatchSnapshot();
  });

  test("response if no matching user", async () => {
    const res = await graphqlTestCall(CONFIRM_LOGIN_MUTATION, {
      token: "FAKECODE"
    });

    expect(res.data.confirmLogin.code).toBe("DOES_NOT_EXIST");
    expect(res.data.confirmLogin.success).toBe(false);
    expect(res.data.confirmLogin.message).toBe(
      "This token is either invalid or expired!"
    );
  });
});
