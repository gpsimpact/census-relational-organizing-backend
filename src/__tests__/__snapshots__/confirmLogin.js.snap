// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`confirmLoginResolver throws error if logged in 1`] = `
Array [
  [GraphQLError: Not Authorized!],
]
`;
